[TCSACR-448] Add cion doc (#1570)
Samsung/tizen-docs: docs/application/native/guides/app-management/cion.md

> # Cion

Author: It's just APIs to develop communicate applications. I'm not sure to add in toc_all.md and what should I have to add.

> # Cion

Author: it's changed

The Cion API provides functionality to communicate with the other devices.

The main features of the Cion API include:

- Communicating with other applications in server-client style.

  You can communicate with other applications in [server-client](#server_client) style.

- Communicating with other applications in group style.

  You can communicate with other applications in [group](#group) style.

- Supporting TIDL

>   You can generate communication code using [TIDL](#tidl).

Author: it's changed

## Prerequisites

1.  To use the Cion API (in [mobile](../../api/mobile/latest/group__CAPI__CION__MODULE.html), [wearable](../../api/wearable/latest/group__CAPI__CION__MODULE.html) and [iot](../../api/iot-headed/latest/group__CAPI__CION__MODULE.html) applications), the application has to request permission by adding the following privilege to the `tizen-manifest.xml` file:

    ```xml
    <privileges>
       <privilege>http://tizen.org/privilege/d2d.datasharing</privilege>
       <privilege>http://tizen.org/privilege/internet</privilege>
    </privileges>
    ```

> 2.  To use the functions of the Widget Service API, include the `<widget_service.h>` header file in your application:

Author: it's changed to cion

    ```cpp
    #include <cion.h>
    ```

<a name="server_client"></a>
## Communicate in server-client style

Cion API provides functions to communicate with the other applications in server-client style.

> Cion client can connect to discovered server and send data.

Reviewer: Global comment: 

Can we add some reference text for procedural steps?

Author: it's changed

> 1. Creates client handle:

Author: it's changed

    ```cpp
    int ret;
    cion_client_h client;

	ret = cion_client_create(&client, "TEST_SERVICE", NULL);
    ```

> 2. Tries to discovery server:

Author: it's changed

    ```cpp
    static void __client_server_discovered_cb(const char *service_name,
		const cion_peer_info_h peer_info, void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_client_try_discovery(client, __client_server_discovered_cb, NULL);
    ```

3. Connects to server:

    ```cpp
    ret = cion_client_connect(client, peer_info);
    ```

    > [!NOTE]
    > You can get the 'peer_info' from discovered callback function.

    > [!NOTE]

>     > You can add callback function using 'cion_client_add_connection_result_cb' to get the connection result.

Author: it's changed

4. Sends data:

    ```cpp
    unsigned char str[] = "test_data";
    unsigned char *return_data = NULL,
    unsigned int return_size;

    ret = cion_client_send_data(client, strlen((const char*)str), 5000, &return_data, &return_size);
    ```
5. Sends payload:

    ```cpp
    static void __cion_client_payload_async_result_cb(
        const cion_payload_async_result_h result, void *user_data);
    ```
    ```cpp
    cion_payload_h payload;
    unsigned char data[] = "test_payload";

    ret = cion_payload_create(&payload, CION_PAYLOAD_TYPE_DATA);
    ret = cion_payload_set_data(payload, data, sizeof(data));

    ret = cion_client_send_payload_async(client, payload, __cion_client_payload_async_result_cb, NULL);
    ```

Cion server can accept connection from client and receive data.

1. Creates server handle:

    ```cpp
    cion_server_h server;

    ret = cion_server_create(&server, "TEST_SERVICE", "MY_DISPLAY_NAME", NULL);
    ```

> 2. Listens the requests from the client:

Author: it's changed

    ```

<a name="group"></a>
## Communicate in group style

Cion API provides functions to communicate with the other applications in group style.

Cion group can subscribe topic and publish data.

1. Creates group handle:

    ```cpp
    cion_group_h group;

	ret = cion_group_create(&group, "test_topic", NULL);
    ```

2. Subscribes topic:

    ```cpp
    ret = cion_group_subscribe(group);
    ```

3. Adds callback function to receive payload:

    ```cpp
    static void __cion_group_payload_received_cb(const char *topic_name,
        const cion_peer_info_h peer_info, cion_payload_h payload, void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_group_add_payload_received_cb(group, __cion_group_payload_received_cb, NULL);
    ```

4. Publishes data:

    ```cpp
   	cion_payload_h payload;
	unsigned char data[] = "test data";

    ret = cion_payload_create(&payload, CION_PAYLOAD_TYPE_DATA);
    ret = cion_payload_set_data(payload, data, sizeof(data));

    ret = cion_group_publish(group, payload);
    ```

<a name="tidl"></a>
## TIDL support

Cion communication code can be generated using TIDL.

> You can make 'CionTest.tidl' file as follows:

Reviewer: ```suggestion
You can create 'CionTest.tidl' file as follows:
```

Author: it's changed

# Cion

> The Cion API provides functionality to communicate with the other devices.

Author: it's changed

The main features of the Cion API include:

- Communicating with other applications in server-client style.

  You can communicate with other applications in [server-client](#server_client) style.

- Communicating with other applications in group style.

  You can communicate with other applications in [group](#group) style.

- Supporting TIDL

  You can generate communication code using [TIDL](#tidl).

## Prerequisites

> 1.  To use the Cion API (in [mobile](../../api/mobile/latest/group__CAPI__CION__MODULE.html), [wearable](../../api/wearable/latest/group__CAPI__CION__MODULE.html) and [iot](../../api/iot-headed/latest/group__CAPI__CION__MODULE.html) applications), the application has to request permission by adding the following privilege to the `tizen-manifest.xml` file:

Reviewer 2: In case of native and web API, they use their own LNB style. Also the stg seems not be updated yet to apply "latest".

I think you can see what is the LNB style of Native API on public.
 - https://docs.tizen.org/application/native/api/mobile/latest/group__CAPI__CION__MODULE.html
 - https://docs.tizen.org/application/native/api/wearable/latest/group__CAPI__CION__MODULE.html
 - https://docs.tizen.org/application/native/api/iot-headed/latest/group__CAPI__CION__MODULE.html

    ```xml
    <privileges>
       <privilege>http://tizen.org/privilege/d2d.datasharing</privilege>
       <privilege>http://tizen.org/privilege/internet</privilege>
    </privileges>
    ```

2.  To use the functions of the Cion API, include the `<cion.h>` header file in your application:

    ```cpp
    #include <cion.h>
    ```

<a name="server_client"></a>
## Communicate in server-client style

Cion API provides functions to communicate with the other applications in server-client style.

> Cion client can try to discover the server to communicate. If server is discovered, client can try to connect to the server.

Author: it's changed

> After client is connected to server, client can send data and paylaod.

Author: it's changed

Cion client can be used as follows:

1. Create the client handle:

    ```cpp
    int ret;
    cion_client_h client;

	ret = cion_client_create(&client, "TEST_SERVICE", NULL);
    ```

2. Discover the server to connect:

    ```cpp
    static void __client_server_discovered_cb(const char *service_name,
		const cion_peer_info_h peer_info, void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_client_try_discovery(client, __client_server_discovered_cb, NULL);
    ```

3. Connect to the server:

    ```cpp
    ret = cion_client_connect(client, peer_info);
    ```

    > [!NOTE]

>     > You can get the 'peer_info' from discovered callback function.

Reviewer: ```suggestion
    > You can get the `peer_info` from discovered callback function.
```

Author: it's changed

    > [!NOTE]
    > You can add callback function using `cion_client_add_connection_result_cb` to get the connection result.

4. Send the data:

    ```cpp
    unsigned char str[] = "test_data";
    unsigned char *return_data = NULL,
    unsigned int return_size;

    ret = cion_client_send_data(client, strlen((const char*)str), 5000, &return_data, &return_size);
    ```
5. Send the payload:

    ```cpp
    static void __cion_client_payload_async_result_cb(
        const cion_payload_async_result_h result, void *user_data);
    ```
    ```cpp
    cion_payload_h payload;
    unsigned char data[] = "test_payload";

    ret = cion_payload_create(&payload, CION_PAYLOAD_TYPE_DATA);
    ret = cion_payload_set_data(payload, data, sizeof(data));

    ret = cion_client_send_payload_async(client, payload, __cion_client_payload_async_result_cb, NULL);
    ```

> Cion server can listen the request from the client. If a connection request comes from the client,the server can accept connection.

Author: it's changed

> Cion server can listen the request from the client. If a connection request comes from the client, the server can accept connection.

Author: it's changed

> After connection is accepted, the server can receive data and payload.

Author: it's changed

> After connection is accepted,the server can receive data and payload.

Reviewer: ```suggestion
After connection is accepted, the server can receive data and payload.
```

Author: it's added

Cion server can be used as follows:

1. Create the server handle:

    ```cpp
    cion_server_h server;

    ret = cion_server_create(&server, "TEST_SERVICE", "MY_DISPLAY_NAME", NULL);
    ```

> 2. Server is ready to listen the requests from the client:

Reviewer 3: Line 117: An ordered (numbered) list should provide a step or instruction to the reader. I'd recommend rephrasing the sentence. I'm adding a suggestion below:
```suggestion
2. Make sure the server is ready to listen to the requests from the client:
```

Author: sentence is changed

    ```cpp
    static void __cion_server_connection_request_cb(const char *service_name,
        const cion_peer_info_h peer_info, void *user_data) 
    {
    }
    ```
    ```cpp
    int cion_server_listen(server, __cion_server_connection_request_cb, void *user_data);
    ```

3. Accept the connection:

    ```cpp
    ret = cion_server_accept(server, peer_info);
    ```

    > [!NOTE]

>     > You can get the 'peer_info' from connection request callback function.

Reviewer: ```suggestion
    > You can get the `peer_info` from connection request callback function.
```

Author: it's changed

4. Add the callback function to receive data and payload:

    ```cpp
    static void __server_data_received_cb(const char *service_name,
	    const cion_peer_info_h peer_info, const unsigned char *data,
		unsigned int data_size, unsigned char **return_data,
		unsigned int *return_data_size, void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_server_set_data_received_cb(server, __server_data_received_cb, NULL);
    ```

    ```cpp
    static void __server_payload_received_cb(const char *service_name,
		const cion_peer_info_h peer_info, const cion_payload_h payload,
		cion_payload_transfer_status_e status,
		void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_server_add_payload_received_cb(server, __server_payload_received_cb, NULL);
    ```

<a name="group"></a>
## Communicate in group style

Cion API provides functions to communicate with the other applications in group style.

> Cion group can subscribe topic and publish data.

Reviewer 3: Line 169: The meaning is not clear, especially the part "...can subscribe topic and publish data"
@sukhyungkang would you please rephrase the sentence? Alternatively, you may explain/elaborate more on what you are trying to say here so I may offer a suggestion...

Author: it's changed

1. Create the group handle:

    ```cpp
    cion_group_h group;

	ret = cion_group_create(&group, "test_topic", NULL);
    ```

2. Subscribe the topic:

    ```cpp
    ret = cion_group_subscribe(group);
    ```

3. Add callback function to receive payload:

    ```cpp
    static void __cion_group_payload_received_cb(const char *topic_name,
        const cion_peer_info_h peer_info, cion_payload_h payload, void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_group_add_payload_received_cb(group, __cion_group_payload_received_cb, NULL);
    ```

> 4. Publishe the data:

Author: it's fixed

    ```cpp
   	cion_payload_h payload;
	unsigned char data[] = "test data";

    ret = cion_payload_create(&payload, CION_PAYLOAD_TYPE_DATA);
    ret = cion_payload_set_data(payload, data, sizeof(data));

    ret = cion_group_publish(group, payload);
    ```

<a name="tidl"></a>
## TIDL support

Cion communication code can be generated using TIDL.

> You can create 'CionTest.tidl' file as follows:

Reviewer: ```suggestion
You can create `CionTest.tidl` file as follows:
```

Author: it's changed

  You can communicate with other applications in [server-client](#server_client) style.

- Communicating with other applications in group style.

  You can communicate with other applications in [group](#group) style.

- Supporting TIDL

  You can generate communication code using [TIDL](#tidl).

## Prerequisites

1.  To use the Cion API (in [mobile](../../api/mobile/latest/group__CAPI__CION__MODULE.html), [wearable](../../api/wearable/latest/group__CAPI__CION__MODULE.html) and [iot](../../api/iot-headed/latest/group__CAPI__CION__MODULE.html) applications), the application has to request permission by adding the following privilege to the `tizen-manifest.xml` file:

    ```xml
    <privileges>
       <privilege>http://tizen.org/privilege/d2d.datasharing</privilege>
       <privilege>http://tizen.org/privilege/internet</privilege>
    </privileges>
    ```

2.  To use the functions of the Cion API, include the `<cion.h>` header file in your application:

    ```cpp
    #include <cion.h>
    ```

<a name="server_client"></a>
## Communicate in server-client style

Cion API provides functions to communicate with the other applications in server-client style.

> The Cion client can try to discover the server to communicate; If server is discovered, the client can try to connect to it.

Reviewer 3: Seems like this correction suggestion was missed from the previous review:
```suggestion
The Cion client can try to discover the server to communicate; if the server is discovered, the client can try to connect to it.
```

Author: it's changed

    {
    }
    ```
    ```cpp
    ret = cion_client_try_discovery(client, __client_server_discovered_cb, NULL);
    ```

3. Connect to the server:

    ```cpp
    ret = cion_client_connect(client, peer_info);
    ```

    > [!NOTE]
    > You can get the `peer_info` from discovered callback function.

    > [!NOTE]
    > You can add callback function using `cion_client_add_connection_result_cb` to get the connection result.

4. Send the data:

    ```cpp
    unsigned char str[] = "test_data";
    unsigned char *return_data = NULL,
    unsigned int return_size;

    ret = cion_client_send_data(client, strlen((const char*)str), 5000, &return_data, &return_size);
    ```
5. Send the payload:

    ```cpp
    static void __cion_client_payload_async_result_cb(
        const cion_payload_async_result_h result, void *user_data);
    ```
    ```cpp
    cion_payload_h payload;
    unsigned char data[] = "test_payload";

    ret = cion_payload_create(&payload, CION_PAYLOAD_TYPE_DATA);
    ret = cion_payload_set_data(payload, data, sizeof(data));

    ret = cion_client_send_payload_async(client, payload, __cion_client_payload_async_result_cb, NULL);
    ```

> The Cion server can listen to the request from the client; If a connection request comes from the client, the server can accept it.

Reviewer 3: ```suggestion
The Cion server can listen to the request from the client; if a connection request comes from the client, the server can accept it.
```

Author: it's changed

After the connection is accepted, the server can receive data and payload.

Cion server can be used as follows:

1. Create the server handle:

    ```cpp
    cion_server_h server;

    ret = cion_server_create(&server, "TEST_SERVICE", "MY_DISPLAY_NAME", NULL);
    ```

> 2. The server starts to listen the requests from the client:

Reviewer 3: ```suggestion
2. The server starts to listen to the requests from the client:
```

Author: it's changed

    ```

3. Accept the connection:

    ```cpp
    ret = cion_server_accept(server, peer_info);
    ```

    > [!NOTE]
    > You can get the `peer_info` from connection request callback function.

4. Add the callback function to receive data and payload:

    ```cpp
    static void __server_data_received_cb(const char *service_name,
	    const cion_peer_info_h peer_info, const unsigned char *data,
		unsigned int data_size, unsigned char **return_data,
		unsigned int *return_data_size, void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_server_set_data_received_cb(server, __server_data_received_cb, NULL);
    ```

    ```cpp
    static void __server_payload_received_cb(const char *service_name,
		const cion_peer_info_h peer_info, const cion_payload_h payload,
		cion_payload_transfer_status_e status,
		void *user_data)
    {
    }
    ```
    ```cpp
    ret = cion_server_add_payload_received_cb(server, __server_payload_received_cb, NULL);
    ```

<a name="group"></a>
## Communicate in group style

The Cion API provides functions to communicate with the other applications in group style.

> The Cion group is used to share data to other group members. The group members can subscribe the topic to join the group.

Reviewer 3: ```suggestion
The Cion group is used to share data with other group members. The group members can subscribe to the topic to join the group.
```

Author: it's changed

> If the data is publised, subscribed members receive it.

Reviewer 3: ```suggestion
If the data is published, subscribed members receive it.
```

Author: it's changed
